crates/subspace-farmer/src/single_disk_farm.rs

```diff
@@ -655,6 +655,98 @@ impl ScrubTarget {
     }
 }
 
+struct AllocatedSpaceDistribution {
+    piece_cache_file_size: u64,
+    piece_cache_capacity: u32,
+    plot_file_size: u64,
+    target_sector_count: u16,
+    metadata_file_size: u64,
+}
+
+impl AllocatedSpaceDistribution {
+    fn new(
+        allocated_space: u64,
+        sector_size: u64,
+        cache_percentage: u8,
+        sector_metadata_size: u64,
+    ) -> Result<Self, SingleDiskFarmError> {
+        let single_sector_overhead = sector_size + sector_metadata_size;
+        // Fixed space usage regardless of plot size
+        let fixed_space_usage = RESERVED_PLOT_METADATA
+            + RESERVED_FARM_INFO
+            + Identity::file_size() as u64
+            + KnownPeersManager::file_size(KNOWN_PEERS_CACHE_SIZE) as u64;
+        // Calculate how many sectors can fit
+        let target_sector_count = {
+            let potentially_plottable_space = allocated_space.saturating_sub(fixed_space_usage)
+                / 100
+                * (100 - u64::from(cache_percentage));
+            // Do the rounding to make sure we have exactly as much space as fits whole number of
+            // sectors, account for disk sector size just in case
+            (potentially_plottable_space - DISK_SECTOR_SIZE as u64) / single_sector_overhead
+        };
+
+        if target_sector_count == 0 {
+            let mut single_plot_with_cache_space =
+                single_sector_overhead.div_ceil(100 - u64::from(cache_percentage)) * 100;
+            // Cache must not be empty, ensure it contains at least one element even if
+            // percentage-wise it will use more space
+            if single_plot_with_cache_space - single_sector_overhead
+                < DiskPieceCache::element_size() as u64
+            {
+                single_plot_with_cache_space =
+                    single_sector_overhead + DiskPieceCache::element_size() as u64;
+            }
+
+            return Err(SingleDiskFarmError::InsufficientAllocatedSpace {
+                min_space: fixed_space_usage + single_plot_with_cache_space,
+                allocated_space,
+            });
+        }
+        let plot_file_size = target_sector_count * sector_size;
+        // Align plot file size for disk sector size
+        let plot_file_size =
+            plot_file_size.div_ceil(DISK_SECTOR_SIZE as u64) * DISK_SECTOR_SIZE as u64;
+
+        // Remaining space will be used for caching purposes
+        let piece_cache_capacity = if cache_percentage > 0 {
+            let cache_space = allocated_space
```

Comment:
Is there a "ratio-check" that prevents the underflow in this snippet?

Reply:
The values used here are either derived or checked a few lines above